File: src/main/java/com/createcivilization/create_ore_deposits/foundation/renderer/block/DepositDrillBlockRenderer.java
package com.createcivilization.create_ore_deposits.foundation.renderer.block;

import com.createcivilization.create_ore_deposits.content.CODPartialModels;
import com.createcivilization.create_ore_deposits.content.blocks.entities.custom.DepositDrillBlockEntity;
import com.simibubi.create.AllPartialModels;
import com.simibubi.create.AllSpriteShifts;
import com.simibubi.create.content.contraptions.pulley.AbstractPulleyRenderer;
import com.simibubi.create.content.fluids.hosePulley.HosePulleyBlock;
import dev.engine_room.flywheel.lib.model.baked.PartialModel;
import net.createmod.catnip.render.CachedBuffers;
import net.createmod.catnip.render.SpriteShiftEntry;
import net.createmod.catnip.render.SuperByteBuffer;
import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.core.Direction;

public class DepositDrillBlockRenderer extends AbstractPulleyRenderer<DepositDrillBlockEntity> {

    public DepositDrillBlockRenderer(BlockEntityRendererProvider.Context context) {
        super(context, AllPartialModels.ROPE_HALF, AllPartialModels.ROPE_HALF_MAGNET);
    }

    @Override
    protected Direction.Axis getShaftAxis(DepositDrillBlockEntity be) {
        return be.getBlockState()
                .getValue(HosePulleyBlock.HORIZONTAL_FACING)
                .getClockWise()
                .getAxis();
    }

    @Override
    protected PartialModel getCoil() {
        return AllPartialModels.HOSE_COIL;
    }

    @Override
    protected SpriteShiftEntry getCoilShift() {
        return AllSpriteShifts.HOSE_PULLEY_COIL;
    }

    @Override
    protected SuperByteBuffer renderRope(DepositDrillBlockEntity be) {
        return CachedBuffers.partial(CODPartialModels.HOSE, be.getBlockState());
    }

    @Override
    protected SuperByteBuffer renderMagnet(DepositDrillBlockEntity be) {
        return CachedBuffers.partial(CODPartialModels.DRILL_MAGNET, be.getBlockState());
    }

    @Override
    protected float getOffset(DepositDrillBlockEntity be, float partialTicks) {
        return be.getInterpolatedOffset(partialTicks);
    }

    @Override
    protected boolean isRunning(DepositDrillBlockEntity be) {
        return true;
    }
}

File: src/main/java/com/createcivilization/create_ore_deposits/registry/block/entries/deposit_drill/DepositDrillBlockRenderer.kt
package com.createcivilization.create_ore_deposits.registry.block.entries.deposit_drill

import com.simibubi.create.AllPartialModels
import com.simibubi.create.AllSpriteShifts
import com.simibubi.create.content.contraptions.pulley.AbstractPulleyRenderer

import com.createcivilization.create_ore_deposits.registry.block.entries.deposit_drill.DepositDrillBlockModels.HOSE
import com.createcivilization.create_ore_deposits.registry.block.entries.deposit_drill.DepositDrillBlockModels.HOSE_HALF
import com.createcivilization.create_ore_deposits.registry.block.entries.deposit_drill.DepositDrillBlockModels.DRILL_MAGNET

import dev.engine_room.flywheel.lib.model.baked.PartialModel

import net.createmod.catnip.render.CachedBuffers
import net.createmod.catnip.render.SpriteShiftEntry
import net.createmod.catnip.render.SuperByteBuffer

import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.core.Direction
import net.minecraft.world.level.block.state.properties.BlockStateProperties

class DepositDrillBlockRenderer(
	context: BlockEntityRendererProvider.Context,
) : AbstractPulleyRenderer<DepositDrillBlockEntity>(context, HOSE, HOSE_HALF) {

	protected override fun getShaftAxis(be: DepositDrillBlockEntity): Direction.Axis =
		be.blockState.getValue(BlockStateProperties.HORIZONTAL_FACING).clockWise.axis

	protected override fun getCoil(): PartialModel = AllPartialModels.HOSE_COIL

	protected override fun getCoilShift(): SpriteShiftEntry = AllSpriteShifts.HOSE_PULLEY_COIL

	override fun renderRope(be: DepositDrillBlockEntity): SuperByteBuffer = CachedBuffers.partial(HOSE, be.blockState)

	override fun renderMagnet(be: DepositDrillBlockEntity): SuperByteBuffer = CachedBuffers.partial(DRILL_MAGNET, be.blockState)

	override fun getOffset(be: DepositDrillBlockEntity, partialTicks: Float): Float = be.getInterpolatedOffset(partialTicks)

	override fun isRunning(be: DepositDrillBlockEntity?): Boolean = true
}